### ISSUE TITLE
Regression in HABTM validation / transaction behaviour

### ISSUE THREAD

**Issue by joshpencheon** [rails/rails#34523]
### Steps to reproduce
```ruby
#!/usr/bin/env ruby

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  # gem "activerecord", "4.2.10"   # Both tests pass
  # gem "activerecord", "5.0.7"    # Both tests pass
  # gem "activerecord", "5.1.6"    # Both tests pass
  # gem "activerecord", "5.2.0"    # Both tests pass
  gem "activerecord", "5.2.1"      # Both tests FAIL !

  # GIT BISECT suggests commit 9f7ba1445524621f is the cause:
  # gem "activesupport", path: 'path/to/git/rails/activesupport'
  # gem "activemodel",   path: 'path/to/git/rails/activemodel'
  # gem "activerecord",  path: 'path/to/git/rails/activerecord'

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :assemblies
  create_table :parts
  create_table :steps

  create_table :assemblies_parts, id: false do |t|
    t.belongs_to :assembly
    t.belongs_to :part
  end

  create_table :assemblies_steps, id: false do |t|
    t.belongs_to :assembly
    t.belongs_to :step
  end
end

class Assembly < ActiveRecord::Base
  has_and_belongs_to_many :parts
  has_and_belongs_to_many :steps
end

class Part < ActiveRecord::Base
  has_and_belongs_to_many :assemblies

  validate { errors.add(:base, 'broken') }
end

class Step < ActiveRecord::Base
  has_and_belongs_to_many :assemblies
end

class BugTest < Minitest::Test
  def setup
    # These objects are persisted in advance of the `assembly`:
    @part = Part.new.tap { |p| p.save(validate: false) }
    @step = Step.new.tap { |s| s.save(validate: false) }
  end

  def test_save_without_transaction
    assembly = Assembly.new
    assembly.parts << @part
    assembly.steps << @step

    assembly.save(validate: false)

    # When not validating, one would expect the join rows to insert:
    assert_equal [1, 1], [assembly.parts.count, assembly.steps.count]
  end

  def test_save_with_transaction
    assembly = Assembly.new
    assembly.parts << @part
    assembly.steps << @step

    Assembly.transaction { assembly.save(validate: false) }

    # When not validating, one would expect the join rows to insert.
    # Certainly NOT expecting the join row for the invalid `part` to
    # insert and the join row for the valid `step` not to!
    assert_equal [1, 1], [assembly.parts.count, assembly.steps.count]
  end
end
```

### Expected behavior
When using `save(validate: false)` to persist a record with HABTM relations to pre-existing other records, both the record and join table entries should persist.

The behaviour should be transactional, and independent of any other open transactions.

### Actual behavior
Without an prior open transaction, neither join row is persisted.
With an existing open transaction, the join row for the _invalid_ pre-existing record _does_ persist, but the join row for the _valid_ pre-existing record _does not_.

### System configuration
**Rails version**:

```
✅ 4.2.10
✅ 5.0.7
✅ 5.1.6
✅ 5.2.0
❌ 5.2.1
⚠️ 9f7ba1445
```

These two test cases both pass on all Rails versions from `4.2.10` up to and including `5.2.0`, but fail on `5.2.1`. 

Using the reproduction script to drive a `git bisect`, it would appear the regression was introduced in 9f7ba1445.

**Ruby version**: 2.5.1

**Comment by joshpencheon** [comment]
For reference, these are the failures I get:

```
  1) Failure:
BugTest#test_save_without_transaction [path/to/habtm_repro.rb:75]:
Expected: [1, 1]
  Actual: [0, 0]

  2) Failure:
BugTest#test_save_with_transaction [path/to/habtm_repro.rb:88]:
Expected: [1, 1]
  Actual: [1, 0]

2 runs, 2 assertions, 2 failures, 0 errors, 0 skips
```

...noting in particular the spliced state in `test_save_with_transaction`.

**Comment by kamipo** [comment]
This is fixed by #33673 and backported 3a2e893b89fb5b352ef749a9c4befec323ce24e8.
